src/sentry/utils/kvstore/bigtable.py

```diff
@@ -0,0 +1,287 @@
+
+        try:
+            cell = columns[self.data_column][0]
+        except KeyError:
+            logger.warning("Retrieved row (%r) which does not contain a data column!", row.row_key)
+            return None
+
+        # Check if a TTL column exists for this row. If there is, we can use
+        # the `timestamp` property of the cells to see if we should return the
+        # row or not.
+        if self.ttl_column in columns:
+            # NOTE: If you need the value, it can be unpacked with the struct.
+            if cell.timestamp < timezone.now():
+                return None
+
+        value = cast(bytes, cell.value)
+
+        if self.flags_column in columns:
+            flags = self.Flags(self.flags_struct.unpack(columns[self.flags_column][0].value)[0])
+
+            # Check if there is a compression flag set, if so decompress the value.
+            # XXX: If no compression flags are matched, we unfortunately can't
+            # tell the difference between data written with a compression
+            # strategy that we're not aware of and data that was not compressed
+            # at all, so we just return the data and hope for the best. It is
+            # also possible that multiple compression flags match. We just stop
+            # after the first one matches. Good luck!
+            for compression_flag, strategy in self.compression_strategies.values():
+                if compression_flag in flags:
+                    value = strategy.decompress(value)
+                    break
+
+        return value
+
+    def set(self, key: str, value: bytes, ttl: Optional[timedelta] = None) -> None:
+        # XXX: There is a type mismatch here -- ``direct_row`` expects
+        # ``bytes`` but we are providing it with ``str``.
+        row = self._get_table().direct_row(key)
+        # Call to delete is just a state mutation, and in this case is just
+        # used to clear all columns so the entire row will be replaced.
+        # Otherwise, if an existing row were mutated, and it took up more than
+        # one column, it'd be possible to overwrite beginning columns and still
+        # retain the end ones.
+        row.delete()
+
+        # If we are setting a TTL on this row, we want to set the timestamp of
+        # the cells into the future. This allows our GC policy to delete them
+        # when the time comes. It also allows us to filter the rows on read if
+        # we are past the timestamp to not return. We want to set a ttl column
+        # to the ttl value in the future if we wanted to bump the timestamp and
+        # rewrite a row with a new ttl.
+        ttl = ttl or self.default_ttl
+        if ttl is None:
+            # XXX: If ``automatic_expiry`` is enabled and no TTL (default or
+            # row-level TTL) is provided, this will default to the Bigtable
+            # server timestamp and this row will be immediately evicted per the
+            # garbage collection policy.
+            ts = None
```

Comment:
The more I think about this, the more easily foot-gunned it seems to be given that both TTLs are default and `automatic_expiry` defaults to `True`.